sys/vane/ames.hoon

```diff
@@ -1365,13 +1365,18 @@
       ~&  [%hear-pubs tea sih]
       =/  our=ship  (slav %p i.t.tea)
       =/  her=ship  (slav %p i.t.t.t.tea)
+      =/  gus  (need (~(us go ton.fox) our))
+      =/  diz  (myx:gus her)
       ?:  =(0 life.sih)
-        ~&  %ames-hear-empty-pubs
-        [~ +>.$]
+          ::  this should clear lew.wod.dur.diz because it means
+          ::  we know longer trust that their public key came to
+          ::  us honestly (becuse of a %jael snapshot restore).
+          ::  in practice, that crashes in ++cluy:las:as:go, so
+          ::  we ignore for now.
```

Comment:
What should we do other than crash? Don't we want to drop packets from ships for which we have no trusted key?

Reply:
We do want to drop those packets.  This crashes the actual event that clears `lew`, which is bad because we need to get the later packets to get the public key again.

One thing to be careful about is making sure we don't end up in a situation where we drop the key for our sponsor and consequently don't get the later packets which would bring us back up to date.

Reply:
> My mental model of this was that we did the "forget, load in snapshot state" in a single event, even to subscribers?

Firstly, this crash means we can't even ack this packet, so the "load snapshot" packet just keeps coming again and again.

Secondly, we do forget and load snapshot state in a single event, but the snapshot could be missing recent events.  Currently we snapshot once every 100 events, then on child boot we send the most recent snapshot followed by the later events, and in theory those missing logs could be important to the conversation.  More generally, the parent could revert back to any block, even block 0.  In that case, all current subscribers would hear "revert to the blank snapshot, then I'll send you events as I hear about them".

Reply:
That's.. completely correct. I was holding onto too many assumptions still, but we can't really make any when we rewind!

Reply:
We could set a timer to disassociate the "processing a snapshot" event from the "received a packet containing a snapshot" event. (Breaking up this and other log processing into separate, batched, smaller events is something I'd like to discuss in more detail). But we'll still definitely need special handling for our sponsor here.

I'm comfortable merging this as is. But we should talk through these scenarios soon. (Not to mention how they interact with discontinuity ...)

Reply:
Yeah, I'll go ahead and merge this tomorrow after fixing @Fang-'s cosmetic notes.

Currently, I believe the only known failure mode is if a block that sets the public key (hereafter, k1) for someone you talk to is reorg-ed out of the chain and *not* re-added to the chain, then we could still talk to them under a key that isn't reflected on chain (i.e. k1).  This isn't a very serious vulnerability because such a transaction can only be sent by the actual owner of the ship, so if e.g. ~zod picked up the tx and spammed it to everyone but the ship owner had a change of heart and decided not to add k1 after all, then unless he bumped the life to another key k2, some people might only recognize k1, which presumably he controls anyway.  He could work around this by bumping his life.

Basically, we don't need much in the way of double-spend protection because the rightful owner can always reclaim his belongings by cycling his keys (or submitting a PR to fix this bug! :) )